rules/manager.go

```diff
@@ -523,6 +527,124 @@ func (m *Manager) Update(interval time.Duration, files []string) error {
 	return nil
 }
 
+// RestoreForState restores the 'for' state of the alerts
+// by looking up last ActiveAt from storage.
+func (m *Manager) RestoreForState(ts time.Time) {
+
+	maxtMS := int64(model.TimeFromUnixNano(ts.UnixNano()))
+	// We allow restoration only if alerts were active before some certain time.
+	mint := ts.Add(-m.opts.OutageTolerance)
+	mintMS := int64(model.TimeFromUnixNano(mint.UnixNano()))
+	q, err := m.opts.TSDB.Querier(m.opts.Context, mintMS, maxtMS)
+	if err != nil {
+		level.Error(m.logger).Log("msg", "Failed to get Querier", "err", err)
+	}
+
+	for _, newg := range m.groups {
+		for _, rule := range newg.Rules() {
+			alertRule, ok := rule.(*AlertingRule)
+			if !ok {
+				continue
+			}
+			_, err := alertRule.Eval(m.opts.Context, ts, newg.opts.QueryFunc, newg.opts.ExternalURL)
+			if err != nil {
+				level.Error(m.logger).Log("msg", "Failed to restore 'for' state",
+					labels.AlertName, alertRule.Name(), "stage", "Eval", "err", err)
+				continue
+			}
+
+			alertHoldDuration := alertRule.HoldDuration()
+
+			alertRule.ForEachActiveAlert(func(a *Alert) {
+
+				smpl := alertRule.forStateSample(a, time.Now(), 0)
+				var matchers []*labels.Matcher
+				for _, l := range smpl.Metric {
+					mt, _ := labels.NewMatcher(labels.MatchEqual, l.Name, l.Value)
+					matchers = append(matchers, mt)
+				}
+
+				sset, err := q.Select(nil, matchers...)
+				if err != nil {
+					level.Error(m.logger).Log("msg", "Failed to restore 'for' state",
+						labels.AlertName, alertRule.Name(), "stage", "Select", "err", err)
+					return
+				}
+
+				seriesFound := false
+				var s storage.Series
+				for sset.Next() {
+					if len(sset.At().Labels()) == len(smpl.Metric) {
+						s = sset.At()
+						seriesFound = true
+						break
+					}
+				}
+
+				if !seriesFound {
+					return
+				}
+
+				// Series found for the 'for' state.
+				var v float64
+				it := s.Iterator()
+				for it.Next() {
+					_, v = it.At()
```

Comment:
I still don't see how this can be right. You're not using the timestamp at which the alert last fired.

Reply:
Before I answer this I have 1 doubt,
If the alert was firing when the prometheus went down, should it be firing again just after restoration or should it wait for the grace period?

Reply:
It should still be firing if it's still evaluating to true I'd say, though it if waits for the grace period that's not the end of the world.

Reply:
I had assumed that it should wait for grace period even in that case. If we have to wait for grace period, we actually don't need to know if the alert was firing. With the `ActiveAt`, we can make out when it should be firing, and then handle the grace period according. 

If we want it to fire immediately after restoration, then we might want to use the timestamp of the sample.

I guess we would like it to fire after restoration as you said. I will make the necessary changes.